services/github/github-api-provider.js

```diff
@@ -52,33 +71,70 @@ class GithubApiProvider {
     if (this.withPooling) {
       this.standardTokens.add(tokenString)
       this.searchTokens.add(tokenString)
+      this.graphqlTokens.add(tokenString)
     } else {
       throw Error('When not using a token pool, do not provide tokens')
     }
   }
 
-  updateToken(token, headers) {
+  getRateLimitFromHeaders(headers) {
+    const h = Joi.attempt(headers, headerSchema)
+    return {
+      rateLimit: h['x-ratelimit-limit'],
+      totalUsesRemaining: h['x-ratelimit-remaining'],
+      nextReset: h['x-ratelimit-reset'],
+    }
+  }
+
+  getRateLimitFromBody(body) {
+    const parsedBody = JSON.parse(body)
+    const b = Joi.attempt(parsedBody, bodySchema)
+    return {
+      rateLimit: b.data.rateLimit.limit,
+      totalUsesRemaining: b.data.rateLimit.remaining,
+      nextReset: Date.parse(b.data.rateLimit.resetAt) / 1000,
+    }
+  }
+
+  updateToken({ token, url, res }) {
     let rateLimit, totalUsesRemaining, nextReset
-    try {
-      ;({
-        'x-ratelimit-limit': rateLimit,
-        'x-ratelimit-remaining': totalUsesRemaining,
-        'x-ratelimit-reset': nextReset,
-      } = Joi.attempt(headers, headerSchema))
-    } catch (e) {
-      const logHeaders = {
-        'x-ratelimit-limit': headers['x-ratelimit-limit'],
-        'x-ratelimit-remaining': headers['x-ratelimit-remaining'],
-        'x-ratelimit-reset': headers['x-ratelimit-reset'],
+    if (url.startsWith('/graphql')) {
+      try {
+        ;({
+          rateLimit,
+          totalUsesRemaining,
+          nextReset,
+        } = this.getRateLimitFromBody(res.body))
+      } catch (e) {
+        console.error(
+          `Could not extract rate limit info from response body ${res.body}`
+        )
+        log.error(e)
+        return
+      }
+    } else {
+      try {
+        ;({
+          rateLimit,
+          totalUsesRemaining,
+          nextReset,
+        } = this.getRateLimitFromHeaders(res.headers))
+      } catch (e) {
+        const logHeaders = {
+          'x-ratelimit-limit': res.headers['x-ratelimit-limit'],
+          'x-ratelimit-remaining': res.headers['x-ratelimit-remaining'],
+          'x-ratelimit-reset': res.headers['x-ratelimit-reset'],
+        }
+        console.error(
+          `Invalid GitHub rate limit headers ${JSON.stringify(
+            logHeaders,
+            undefined,
+            2
+          )}`
+        )
+        log.error(e)
```

Comment:
Note that I am changing the behaviour here. Previously we were just swallowing this error in production, but it seems like if we end up here, we should be logging something to sentry so we can go and look at it. Taking this path means we are making calls using tokens but we're not updating the `usesRemaining` value. I assume this isn't happening routinely because we're not maxing out everyone's tokens, but I'd still like to know about it if it does happen.